`cmse-nonsecure-entry` hits query cycle when returning `impl Trait`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

Just documenting this so we have a paper trail. I intend to resolve this by disallowing `impl Trait` in return types altogether: we already disallow standard generics on functions with this ABI.

### Code

```Rust
#![feature(cmse_nonsecure_entry)]
#![no_std]

extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
    0u128
}

#[unsafe(no_mangle)]
fn foo() {
    let _x = return_impl_trait();
}

```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
cargo 1.92.0-nightly (966f94733 2025-09-16)
release: 1.92.0-nightly
commit-hash: 966f94733bbc94ca51ff9f1e4c49ad250ebbdc50
commit-date: 2025-09-16
host: x86_64-unknown-linux-gnu
libgit2: 1.9.1 (sys:0.20.2 vendored)
libcurl: 8.15.0-DEV (sys:0.4.83+curl-8.15.0 vendored ssl:OpenSSL/3.5.2)
ssl: OpenSSL 3.5.2 5 Aug 2025
os: Linux Mint 22.0.0 (wilma) [64-bit]
```

### Error output

```
warning: `extern` fn uses type `impl core::marker::Copy`, which is not FFI-safe
 --> src/lib.rs:4:57
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  |                                                         ^^^^^^^^^^^^^^^^^^^^^^^ not FFI-safe
  |
  = note: opaque types have no C equivalent
  = note: `#[warn(improper_ctypes_definitions)]` on by default

note: no errors encountered even though delayed bugs were created

note: those delayed bugs will now be shown as internal compiler errors

error: internal compiler error[E0391]: cycle detected when computing function signature of `return_impl_trait`
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: ...which requires computing layout of `return_impl_trait::{opaque#0}`...
  = note: ...which requires normalizing `return_impl_trait::{opaque#0}`...
note: ...which requires computing type of `return_impl_trait::{opaque#0}`...
 --> src/lib.rs:4:57
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  |                                                         ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires computing type of opaque `return_impl_trait::{opaque#0}`...
 --> src/lib.rs:4:57
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  |                                                         ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires borrow-checking `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires promoting constants in MIR for `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires checking if `return_impl_trait` contains FFI-unwind calls...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires building MIR for `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires match-checking `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires type-checking `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires computing the opaque types defined by `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: ...which again requires computing function signature of `return_impl_trait`, completing the cycle
note: cycle used when checking that `return_impl_trait` is well-formed
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
note: delayed at compiler/rustc_query_system/src/query/job.rs:602:16 - disabled backtrace
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: internal compiler error[E0391]: cycle detected when computing function signature of `return_impl_trait`
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: ...which requires computing layout of `return_impl_trait::{opaque#0}`...
  = note: ...which requires normalizing `return_impl_trait::{opaque#0}`...
note: ...which requires computing type of `return_impl_trait::{opaque#0}`...
 --> src/lib.rs:4:57
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  |                                                         ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires computing type of opaque `return_impl_trait::{opaque#0}`...
 --> src/lib.rs:4:57
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  |                                                         ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires borrow-checking `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires promoting constants in MIR for `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires checking if `return_impl_trait` contains FFI-unwind calls...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires building MIR for `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires match-checking `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires type-checking `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: ...which again requires computing function signature of `return_impl_trait`, completing the cycle
note: cycle used when checking that `return_impl_trait` is well-formed
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
note: delayed at compiler/rustc_query_system/src/query/job.rs:602:16 - disabled backtrace
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: internal compiler error[E0391]: cycle detected when computing function signature of `return_impl_trait`
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: ...which requires computing layout of `return_impl_trait::{opaque#0}`...
  = note: ...which requires normalizing `return_impl_trait::{opaque#0}`...
note: ...which requires computing type of `return_impl_trait::{opaque#0}`...
 --> src/lib.rs:4:57
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  |                                                         ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires computing type of opaque `return_impl_trait::{opaque#0}`...
 --> src/lib.rs:4:57
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  |                                                         ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires borrow-checking `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires promoting constants in MIR for `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires checking if `return_impl_trait` contains FFI-unwind calls...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires building MIR for `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: ...which again requires computing function signature of `return_impl_trait`, completing the cycle
note: cycle used when checking that `return_impl_trait` is well-formed
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
note: delayed at compiler/rustc_query_system/src/query/job.rs:602:16 - disabled backtrace
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: internal compiler error[E0391]: cycle detected when computing function signature of `return_impl_trait`
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: ...which requires computing layout of `return_impl_trait::{opaque#0}`...
  = note: ...which requires normalizing `return_impl_trait::{opaque#0}`...
note: ...which requires computing type of `return_impl_trait::{opaque#0}`...
 --> src/lib.rs:4:57
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  |                                                         ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires computing type of opaque `return_impl_trait::{opaque#0}`...
 --> src/lib.rs:4:57
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  |                                                         ^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires borrow-checking `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires promoting constants in MIR for `return_impl_trait`...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: ...which requires checking if `return_impl_trait` contains FFI-unwind calls...
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: ...which again requires computing function signature of `return_impl_trait`, completing the cycle
note: cycle used when checking that `return_impl_trait` is well-formed
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: see https://rustc-dev-guide.rust-lang.org/overview.html#queries and https://rustc-dev-guide.rust-lang.org/query.html for more information
note: delayed at compiler/rustc_query_system/src/query/job.rs:602:16 - disabled backtrace
 --> src/lib.rs:4:1
  |
4 | extern "cmse-nonsecure-entry" fn return_impl_trait() -> impl core::marker::Copy {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/home/folkertdev/rust/playground/rustc-ice-2025-10-01T18_43_37-135196.txt` to your bug report

note: compiler flags: --crate-type lib -C opt-level=3 -C panic=abort -C embed-bitcode=no -C debuginfo=2 -Z unstable-options -C llvm-args=-x86-asm-syntax=intel -C debuginfo=2 -C codegen-units=1

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
warning: `playground` (lib) generated 1 warning
error: could not compile `playground` (lib); 1 warning emitted

Caused by:
  process didn't exit successfully: `/home/folkertdev/.rustup/toolchains/stage1/bin/rustc --crate-name playground --edition=2021 src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=190 --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C panic=abort -C embed-bitcode=no -C debuginfo=2 --check-cfg 'cfg(docsrs,test)' --check-cfg 'cfg(feature, values())' -C metadata=f3bf8d08de02bd06 -C extra-filename=-642b5b0b4127d905 --out-dir /home/folkertdev/rust/playground/target/thumbv8m.main-none-eabi/release/deps --target thumbv8m.main-none-eabi -L dependency=/home/folkertdev/rust/playground/target/thumbv8m.main-none-eabi/release/deps -L dependency=/home/folkertdev/rust/playground/target/release/deps --extern 'noprelude:compiler_builtins=/home/folkertdev/rust/playground/target/thumbv8m.main-none-eabi/release/deps/libcompiler_builtins-eb37b01561e12bc2.rmeta' --extern 'noprelude:core=/home/folkertdev/rust/playground/target/thumbv8m.main-none-eabi/release/deps/libcore-5dcd424cf93dc313.rmeta' -Z unstable-options --emit asm -C llvm-args=-x86-asm-syntax=intel -Cdebuginfo=2 -Ccodegen-units=1` (exit status: 101)
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
<backtrace>
```

</p>
</details>

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/cmse.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/cmse.rs
@@ -3,7 +3,7 @@ use rustc_errors::{DiagCtxtHandle, E0781, struct_span_code_err};
 use rustc_hir::{self as hir, HirId};
 use rustc_middle::bug;
 use rustc_middle::ty::layout::{LayoutError, TyAndLayout};
-use rustc_middle::ty::{self, TyCtxt};
+use rustc_middle::ty::{self, TyCtxt, TypeVisitableExt};
 
 use crate::errors;
 
@@ -165,9 +165,23 @@ fn is_valid_cmse_output<'tcx>(
     // this type is only used for layout computation, which does not rely on regions
     let fn_sig = tcx.instantiate_bound_regions_with_erased(fn_sig);
     let fn_sig = tcx.erase_and_anonymize_regions(fn_sig);
+    let return_type = fn_sig.output();
+
+    // `impl Trait` is already disallowed with `cmse-nonsecure-call`, because that ABI is only
+    // allowed on function pointers, and function pointers cannot contain `impl Trait` in their
+    // signature.
+    //
+    // Here we explicitly disallow `impl Trait` in the `cmse-nonsecure-entry` return type too, to
+    // prevent query cycles when calculating the layout. This ABI is meant to be used with
+    // `#[no_mangle]` or similar, so generics in the type really don't make sense.
+    //
+    // see also https://github.com/rust-lang/rust/issues/147242.
+    if return_type.has_opaque_types() {
+        return Err(tcx.arena.alloc(LayoutError::TooGeneric(return_type)));
+    }
 
     let typing_env = ty::TypingEnv::fully_monomorphized();
-    let layout = tcx.layout_of(typing_env.as_query_input(fn_sig.output()))?;
+    let layout = tcx.layout_of(typing_env.as_query_input(return_type))?;
 
     Ok(is_valid_cmse_output_layout(layout))
 }

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/cmse.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/cmse.rs
@@ -1,8 +1,8 @@
-use rustc_abi::ExternAbi;
+use rustc_abi::{BackendRepr, ExternAbi, Float, Integer, Primitive, Scalar};
 use rustc_errors::{DiagCtxtHandle, E0781, struct_span_code_err};
 use rustc_hir::{self as hir, HirId};
 use rustc_middle::bug;
-use rustc_middle::ty::layout::LayoutError;
+use rustc_middle::ty::layout::{LayoutError, TyAndLayout};
 use rustc_middle::ty::{self, TyCtxt};
 
 use crate::errors;
@@ -164,44 +164,34 @@ fn is_valid_cmse_output<'tcx>(
 ) -> Result<bool, &'tcx LayoutError<'tcx>> {
     // this type is only used for layout computation, which does not rely on regions
     let fn_sig = tcx.instantiate_bound_regions_with_erased(fn_sig);
+    let fn_sig = tcx.erase_and_anonymize_regions(fn_sig);
 
     let typing_env = ty::TypingEnv::fully_monomorphized();
+    let layout = tcx.layout_of(typing_env.as_query_input(fn_sig.output()))?;
 
-    let mut ret_ty = fn_sig.output();
-    let layout = tcx.layout_of(typing_env.as_query_input(ret_ty))?;
+    Ok(is_valid_cmse_output_layout(layout))
+}
+
+/// Returns whether the output will fit into the available registers
+fn is_valid_cmse_output_layout<'tcx>(layout: TyAndLayout<'tcx>) -> bool {
     let size = layout.layout.size().bytes();
 
     if size <= 4 {
-        return Ok(true);
+        return true;
     } else if size > 8 {
-        return Ok(false);
+        return false;
     }
 
-    // next we need to peel any repr(transparent) layers off
-    'outer: loop {
-        let ty::Adt(adt_def, args) = ret_ty.kind() else {
-            break;
-        };
+    // Accept scalar 64-bit types.
+    let BackendRepr::Scalar(scalar) = layout.layout.backend_repr else {
+        return false;
+    };
 
-        if !adt_def.repr().transparent() {
-            break;
-        }
+    let Scalar::Initialized { value, .. } = scalar else {
+        return false;
+    };
 
-        // the first field with non-trivial size and alignment must be the data
-        for variant_def in adt_def.variants() {
-            for field_def in variant_def.fields.iter() {
-                let ty = field_def.ty(tcx, args);
-                let layout = tcx.layout_of(typing_env.as_query_input(ty))?;
-
-                if !layout.layout.is_1zst() {
-                    ret_ty = ty;
-                    continue 'outer;
-                }
-            }
-        }
-    }
-
-    Ok(ret_ty == tcx.types.i64 || ret_ty == tcx.types.u64 || ret_ty == tcx.types.f64)
+    matches!(value, Primitive::Int(Integer::I64, _) | Primitive::Float(Float::F64))
 }
 
 fn should_emit_generic_error<'tcx>(abi: ExternAbi, layout_err: &'tcx LayoutError<'tcx>) -> bool {

--- a/tests/ui/cmse-nonsecure/cmse-nonsecure-call/generics.rs
+++ b/tests/ui/cmse-nonsecure/cmse-nonsecure-call/generics.rs
@@ -14,8 +14,9 @@ struct Test<T: Copy> {
     f1: extern "cmse-nonsecure-call" fn<U: Copy>(U, u32, u32, u32) -> u64,
     //~^ ERROR cannot find type `U` in this scope
     //~| ERROR function pointer types may not have generic parameters
-    f2: extern "cmse-nonsecure-call" fn(impl Copy, u32, u32, u32) -> u64,
+    f2: extern "cmse-nonsecure-call" fn(impl Copy, u32, u32, u32) -> impl Copy,
     //~^ ERROR `impl Trait` is not allowed in `fn` pointer parameters
+    //~| ERROR `impl Trait` is not allowed in `fn` pointer return types
     f3: extern "cmse-nonsecure-call" fn(T, u32, u32, u32) -> u64, //~ ERROR [E0798]
     f4: extern "cmse-nonsecure-call" fn(Wrapper<T>, u32, u32, u32) -> u64, //~ ERROR [E0798]
 }

--- a/tests/ui/cmse-nonsecure/cmse-nonsecure-call/generics.stderr
+++ b/tests/ui/cmse-nonsecure/cmse-nonsecure-call/generics.stderr
@@ -25,25 +25,33 @@ LL | struct Test<T: Copy, U> {
 error[E0562]: `impl Trait` is not allowed in `fn` pointer parameters
   --> $DIR/generics.rs:17:41
    |
-LL |     f2: extern "cmse-nonsecure-call" fn(impl Copy, u32, u32, u32) -> u64,
+LL |     f2: extern "cmse-nonsecure-call" fn(impl Copy, u32, u32, u32) -> impl Copy,
    |                                         ^^^^^^^^^
    |
    = note: `impl Trait` is only allowed in arguments and return types of functions and methods
 
+error[E0562]: `impl Trait` is not allowed in `fn` pointer return types
+  --> $DIR/generics.rs:17:70
+   |
+LL |     f2: extern "cmse-nonsecure-call" fn(impl Copy, u32, u32, u32) -> impl Copy,
+   |                                                                      ^^^^^^^^^
+   |
+   = note: `impl Trait` is only allowed in arguments and return types of functions and methods
+
 error[E0798]: function pointers with the `"cmse-nonsecure-call"` ABI cannot contain generics in their type
-  --> $DIR/generics.rs:19:9
+  --> $DIR/generics.rs:20:9
    |
 LL |     f3: extern "cmse-nonsecure-call" fn(T, u32, u32, u32) -> u64,
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error[E0798]: function pointers with the `"cmse-nonsecure-call"` ABI cannot contain generics in their type
-  --> $DIR/generics.rs:20:9
+  --> $DIR/generics.rs:21:9
    |
 LL |     f4: extern "cmse-nonsecure-call" fn(Wrapper<T>, u32, u32, u32) -> u64,
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error[E0798]: return value of `"cmse-nonsecure-call"` function too large to pass via registers
-  --> $DIR/generics.rs:26:71
+  --> $DIR/generics.rs:27:71
    |
 LL | type WithTraitObject = extern "cmse-nonsecure-call" fn(&dyn Trait) -> &dyn Trait;
    |                                                                       ^^^^^^^^^^ this type doesn't fit in the available registers
@@ -52,7 +60,7 @@ LL | type WithTraitObject = extern "cmse-nonsecure-call" fn(&dyn Trait) -> &dyn 
    = note: the result must either be a (transparently wrapped) i64, u64 or f64, or be at most 4 bytes in size
 
 error[E0798]: return value of `"cmse-nonsecure-call"` function too large to pass via registers
-  --> $DIR/generics.rs:30:60
+  --> $DIR/generics.rs:31:60
    |
 LL |     extern "cmse-nonsecure-call" fn(&'static dyn Trait) -> &'static dyn Trait;
    |                                                            ^^^^^^^^^^^^^^^^^^ this type doesn't fit in the available registers
@@ -61,7 +69,7 @@ LL |     extern "cmse-nonsecure-call" fn(&'static dyn Trait) -> &'static dyn Tra
    = note: the result must either be a (transparently wrapped) i64, u64 or f64, or be at most 4 bytes in size
 
 error[E0798]: return value of `"cmse-nonsecure-call"` function too large to pass via registers
-  --> $DIR/generics.rs:37:60
+  --> $DIR/generics.rs:38:60
    |
 LL |     extern "cmse-nonsecure-call" fn(WrapperTransparent) -> WrapperTransparent;
    |                                                            ^^^^^^^^^^^^^^^^^^ this type doesn't fit in the available registers
@@ -70,12 +78,12 @@ LL |     extern "cmse-nonsecure-call" fn(WrapperTransparent) -> WrapperTranspare
    = note: the result must either be a (transparently wrapped) i64, u64 or f64, or be at most 4 bytes in size
 
 error[E0045]: C-variadic functions with the "cmse-nonsecure-call" calling convention are not supported
-  --> $DIR/generics.rs:40:20
+  --> $DIR/generics.rs:41:20
    |
 LL | type WithVarArgs = extern "cmse-nonsecure-call" fn(u32, ...);
    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ C-variadic function must have a compatible calling convention
 
-error: aborting due to 9 previous errors
+error: aborting due to 10 previous errors
 
 Some errors have detailed explanations: E0045, E0412, E0562, E0798.
 For more information about an error, try `rustc --explain E0045`.

--- a/tests/ui/cmse-nonsecure/cmse-nonsecure-entry/generics.rs
+++ b/tests/ui/cmse-nonsecure/cmse-nonsecure-entry/generics.rs
@@ -65,3 +65,15 @@ extern "cmse-nonsecure-entry" fn wrapped_trait_object(x: WrapperTransparent) -> 
     //~^ ERROR return value of `"cmse-nonsecure-entry"` function too large to pass via registers [E0798]
     x
 }
+
+extern "cmse-nonsecure-entry" fn return_impl_trait(_: impl Copy) -> impl Copy {
+    //~^ ERROR functions with the `"cmse-nonsecure-entry"` ABI cannot contain generics in their type
+    //~| ERROR functions with the `"cmse-nonsecure-entry"` ABI cannot contain generics in their type
+    0u128
+}
+
+extern "cmse-nonsecure-entry" fn return_impl_trait_nested(v: (impl Copy, i32)) -> (impl Copy, i32) {
+    //~^ ERROR functions with the `"cmse-nonsecure-entry"` ABI cannot contain generics in their type
+    //~| ERROR functions with the `"cmse-nonsecure-entry"` ABI cannot contain generics in their type
+    v
+}

--- a/tests/ui/cmse-nonsecure/cmse-nonsecure-entry/generics.stderr
+++ b/tests/ui/cmse-nonsecure/cmse-nonsecure-entry/generics.stderr
@@ -17,6 +17,34 @@ LL | extern "cmse-nonsecure-entry" fn impl_trait(_: impl Copy, _: u32, _: u32, _
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 error[E0798]: functions with the `"cmse-nonsecure-entry"` ABI cannot contain generics in their type
+  --> $DIR/generics.rs:69:1
+   |
+LL | extern "cmse-nonsecure-entry" fn return_impl_trait(_: impl Copy) -> impl Copy {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error[E0798]: functions with the `"cmse-nonsecure-entry"` ABI cannot contain generics in their type
+  --> $DIR/generics.rs:69:1
+   |
+LL | extern "cmse-nonsecure-entry" fn return_impl_trait(_: impl Copy) -> impl Copy {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+
+error[E0798]: functions with the `"cmse-nonsecure-entry"` ABI cannot contain generics in their type
+  --> $DIR/generics.rs:75:1
+   |
+LL | extern "cmse-nonsecure-entry" fn return_impl_trait_nested(v: (impl Copy, i32)) -> (impl Copy, i32) {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+error[E0798]: functions with the `"cmse-nonsecure-entry"` ABI cannot contain generics in their type
+  --> $DIR/generics.rs:75:1
+   |
+LL | extern "cmse-nonsecure-entry" fn return_impl_trait_nested(v: (impl Copy, i32)) -> (impl Copy, i32) {
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
+
+error[E0798]: functions with the `"cmse-nonsecure-entry"` ABI cannot contain generics in their type
   --> $DIR/generics.rs:14:5
    |
 LL |     extern "cmse-nonsecure-entry" fn ambient_generic(_: T, _: u32, _: u32, _: u32) -> u64 {
@@ -61,6 +89,6 @@ LL | extern "cmse-nonsecure-entry" fn wrapped_trait_object(x: WrapperTransparent
    = note: functions with the `"cmse-nonsecure-entry"` ABI must pass their result via the available return registers
    = note: the result must either be a (transparently wrapped) i64, u64 or f64, or be at most 4 bytes in size
 
-error: aborting due to 7 previous errors
+error: aborting due to 11 previous errors
 
 For more information about this error, try `rustc --explain E0798`.

